### PR TITLE
Update in MFCK/ SMFCK

### PR DESCRIPTION
**Pull request by mcastanoUQ**
**Summary**

Extends the MFCK and SMFCK multi-fidelity surrogate models with gradient-based
hyperparameter optimization, heteroscedastic noise prediction, and a sequential
optimization scheme. Also includes an internal refactoring that makes the shared
co-Kriging machinery reusable and inspectable.

**Modularization**

Shared logic previously duplicated across the MFCK and SMFCK implementations has
been extracted into helper methods:

- Cholesky decomposition of the covariance matrix
- Cross-covariance assembly between fidelity levels
- Diagonal prior covariance construction

Beyond removing duplication, this exposes the intermediate quantities so that
internal behaviour (conditioning, cross-level correlations) can be inspected
directly rather than only through the final prediction.

**Gradients for MFCK**

Analytical gradients of the log marginal likelihood with respect to the kernel
hyperparameters are now implemented for the RBF kernel. This allows MFCK to be
trained with gradient-based optimizers such as TNC.

**Heteroscedastic noise prediction**

Both MFCK and SMFCK can now predict input-dependent noise

**Sequential optimization**
An option that integrates hyperparameter optimization level by level rather
than jointly. This option reduces the training complexity of MFCK to make it comparable to that of MFK.

## Notebooks

Updated to use the sequential optimization option in all tutorials associated:

- `tutorial/MultiFi/SMT_MFCK_tutorial.ipynb`
- `tutorial/MultiFi/SMT_SMFCK_tutorial.ipynb`
- `tutorial/MultiFi/SMT_SMFK_tutorial.ipynb`
- `tutorial/MultiFi/Paper_SMFCK_Notebook_Rosenbrock.ipynb`

### CODE REVIEW

File: smt/applications/smfk.py [hunk 1]
```diff
@@ -396,9 +396,14 @@ def predict_variances_all_levels(self, X, is_acting=None):
                     + sigma2 * (u_**2).sum(axis=0)
                 )
             else:
+                print(
```
Comment (relf) [inline smt/applications/smfk.py]: print statement should be removed

File: smt/applications/smfk.py [hunk 2]
```diff
@@ -396,9 +396,14 @@ def predict_variances_all_levels(self, X, is_acting=None):
...
+                    self.optimal_noise_all[i].shape,
+                    (r_t**2).sum(axis=0).shape,
+                    (u_**2).sum(axis=0).shape,
+                )
                 MSE[:, i] = sigma2 * (
                     1
-                    + self.optimal_noise_all[i]
+                    # + self.optimal_noise_all[i]
```
Comment (relf) [inline smt/applications/smfk.py:401]: to be removed or ... ?

**Review by relf — APPROVED** [review]
Thanks. Minor comments to address before merging

File: smt/applications/smfk.py (hunk 1 above)
Comment (mcastanoUQ) [reply smt/applications/smfk.py]: Removing it

File: smt/applications/smfk.py (hunk 2 above)
Comment (mcastanoUQ) [reply smt/applications/smfk.py:401]: Yes, now it presents an inconsistency with the shapes. I need to check that; for the moment, we can remove it. That line is to take into account the noise provided by the experts when working with het. noise

Comment (relf) [reply smt/applications/smfk.py:401]: Ok so just update the PR description to mention that smfk does not take into account the noise
